test/solidity/Periphery/ServiceFeeCollector.t.sol

```diff
@@ -0,0 +1,179 @@
+            address(0xb33f)
+        );
+        uint256 startingBalance = feeToken.balanceOf(address(this));
+
+        // Act
+        feeCollector.withdrawFees(address(feeToken));
+
+        // Assert
+        assert(
+            feeToken.balanceOf(address(this)) == 0.015 ether + startingBalance
+        );
+        assert(feeToken.balanceOf(address(feeCollector)) == 0);
+    }
+
+    function testCanBatchWithdrawGasFees() public {
+        // Arrange
+        uint256 fee = 0.015 ether;
+        feeToken.approve(address(feeCollector), fee);
+        feeCollector.collectTokenGasFees(
+            address(feeToken),
+            fee,
+            address(0xb33f)
+        );
+        feeCollector.collectNativeGasFees{ value: fee }(fee, address(0xb33f));
+        uint256 startingTokenBalance = feeToken.balanceOf(address(this));
+        uint256 startingETHBalance = address(this).balance;
+
+        // Act
+        address[] memory tokens = new address[](2);
+        tokens[0] = address(feeToken);
+        tokens[1] = address(0);
+        feeCollector.batchWithdrawFees(tokens);
+
+        // Assert
+        assert(
+            feeToken.balanceOf(address(this)) ==
+                0.015 ether + startingTokenBalance
+        );
+        assert(address(this).balance == 0.015 ether + startingETHBalance);
+        assert(address(feeCollector).balance == 0);
+    }
+
+    function testFailWhenNonOwnerAttemptsToWithdrawFees() public {
+        // Arrange
+        uint256 fee = 0.015 ether;
+        feeToken.approve(address(feeCollector), fee);
+        feeCollector.collectTokenGasFees(
+            address(feeToken),
+            fee,
+            address(0xb33f)
+        );
+
+        // Act
+        vm.prank(address(0xb33f));
+        feeCollector.withdrawFees(address(feeToken));
+    }
+
+    function testOwnerCanTransferOwnership() public {
+        address newOwner = address(0x1234567890123456789012345678901234567890);
+        feeCollector.transferOwnership(newOwner);
+        assert(feeCollector.owner() != newOwner);
+        vm.startPrank(newOwner);
+        feeCollector.confirmOwnershipTransfer();
+        assert(feeCollector.owner() == newOwner);
+        vm.stopPrank();
+    }
+
+    function testFailNonOwnerCanTransferOwnership() public {
+        address newOwner = address(0x1234567890123456789012345678901234567890);
+        assert(feeCollector.owner() != newOwner);
+        vm.prank(newOwner);
+        feeCollector.transferOwnership(newOwner);
+    }
+
+    function testFailOnwershipTransferToNullAddr() public {
+        address newOwner = address(0x0);
+        feeCollector.transferOwnership(newOwner);
+    }
+
+    function testFailOwnerCanConfirmPendingOwnershipTransfer() public {
+        address newOwner = address(0x1234567890123456789012345678901234567890);
+        feeCollector.transferOwnership(newOwner);
+        feeCollector.confirmOwnershipTransfer();
+    }
+
+    function testFailOwnershipTransferToSelf() public {
```

Comment:
I would suggest that in cases where we know which error to expect, we should always catch the exact revert reason and make sure that the tests fail for the reason we expect (example in TestBase). I had situations before where an issue slipped through my tests because of that.

Reply:
Good call.